src/vcpkg/base/system.process.cpp

```diff
@@ -501,6 +502,23 @@ namespace vcpkg
         {
             w.get();
         }
+
+#if defined(__APPLE__)
```

Comment:
Note that @ras0219-msft Didn't want to have this inside `cmd_execute_and_stream_data`: https://github.com/microsoft/vcpkg-tool/pull/306#issuecomment-1007031965

Because this is an issue for every command that is executed in parallel, I think it is more reasonable to put it here than just putting it in `decompress_in_parallel`

Reply:
Why is this an issue for every command executed in parallel? I am strongly against this change without a thorough understanding of why it happens and why it's safe to rerun any arbitrary process with a 127 exit code.

Reply:
> Why is this an issue for every command executed in parallel?

I really don't know. I guess this is something you have to ask the folks at Apple since this is most likely not our fault. I read the documentation for every POSIX function that is involved and couldn't find anything. Therefore, I think this is an internal bug of `popen` / `pclose`.

> without a thorough understanding of why it happens

The issue seems to occur only when many processes are launched at the same time with `popen` but this is just guessing. You better ask the Apple folks as we can't step into this.

> why it's safe to rerun any arbitrary process with a 127 exit code.

We're not running every process again that exits with 127 - only if there is no error message (console output). I'm pretty sure there is an error message if you did `eho` instead of `echo` and thus was a real 127. 127 may also be returned if the system is unable to execute the command although it exists: https://www.ibm.com/support/pages/exit-code-127-means-jobs-command-can-not-be-found-or-executed

What else do you expect to happen when you rerun a program that returned 127?

Reply:
Doing further research, I think it's likely due to opening and closing files internally as part of `popen()`. The full process is perhaps not thread safe because it relies on file descriptor inheritance / numbering.

Instead of continuing to try to patch over race conditions, I think we should fix this problem correctly in one of two ways:
1. (cheap and easy) Add a mutex around the `popen()` call on MacOS.
2. (more involved) Switch to using `posix_spawn()` for MacOS.

Reply:
I will go for a static mutex inside the function that uses `popen` instead of passing it around as a pointer because this seems to be the least messy way.

Reply:
address sanitizer also found this:
```
AddressSanitizer:DEADLYSIGNAL
=================================================================
==7417==ERROR: AddressSanitizer: BUS on unknown address (pc 0x00018ad7c2bc bp 0x00016d02d0d0 sp 0x00016d02d050 T407)
==7417==The signal is caused by a UNKNOWN memory access.
==7417==Hint: this fault was caused by a dereference of a high value address (see register values below).  Disassemble the provided pc to learn which register was used.
AddressSanitizer:DEADLYSIGNAL
    #0 0x18ad7c2bc in popen+0x230 (libsystem_c.dylib:arm64e+0x522bc)
    #1 0x107dcfe4c in wrap_popen+0x2ac (libclang_rt.asan_osx_dynamic.dylib:arm64e+0x37e4c)
    #2 0x103308a94 in vcpkg::cmd_execute_and_stream_data(vcpkg::Command const&, std::__1::function<void (vcpkg::StringView)>, vcpkg::WorkingDirectory const&, vcpkg::Environment const&, vcpkg::Encoding) system.process.cpp:927
    #3 0x1032fffe8 in vcpkg::cmd_execute_and_capture_output(vcpkg::Command const&, vcpkg::WorkingDirectory const&, vcpkg::Environment const&, vcpkg::Encoding, vcpkg::EchoInDebug) system.process.cpp:984
    #4 0x103301c44 in vcpkg::cmd_execute_and_capture_output_parallel(vcpkg::Span<vcpkg::Command const>, vcpkg::WorkingDirectory const&, vcpkg::Environment const&)::$_2::operator()() const system.process.cpp:489
    #5 0x103327f18 in decltype(static_cast<vcpkg::cmd_execute_and_capture_output_parallel(vcpkg::Span<vcpkg::Command const>, vcpkg::WorkingDirectory const&, vcpkg::Environment const&)::$_2>(fp)()) std::__1::__invoke<vcpkg::cmd_execute_and_capture_output_parallel(vcpkg::Span<vcpkg::Command const>, vcpkg::WorkingDirectory const&, vcpkg::Environment const&)::$_2>(vcpkg::cmd_execute_and_capture_output_parallel(vcpkg::Span<vcpkg::Command const>, vcpkg::WorkingDirectory const&, vcpkg::Environment const&)::$_2&&) type_traits:3918
    #6 0x103327ebc in void std::__1::__async_func<vcpkg::cmd_execute_and_capture_output_parallel(vcpkg::Span<vcpkg::Command const>, vcpkg::WorkingDirectory const&, vcpkg::Environment const&)::$_2>::__execute<>(std::__1::__tuple_indices<>) future:2190
    #7 0x103327db4 in std::__1::__async_func<vcpkg::cmd_execute_and_capture_output_parallel(vcpkg::Span<vcpkg::Command const>, vcpkg::WorkingDirectory const&, vcpkg::Environment const&)::$_2>::operator()() future:2183
    #8 0x1033274e8 in std::__1::__async_assoc_state<void, std::__1::__async_func<vcpkg::cmd_execute_and_capture_output_parallel(vcpkg::Span<vcpkg::Command const>, vcpkg::WorkingDirectory const&, vcpkg::Environment const&)::$_2> >::__execute() future:1001
    #9 0x10332c678 in decltype(*(static_cast<std::__1::__async_assoc_state<void, std::__1::__async_func<vcpkg::cmd_execute_and_capture_output_parallel(vcpkg::Span<vcpkg::Command const>, vcpkg::WorkingDirectory const&, vcpkg::Environment const&)::$_2> >*>(fp0)).*fp()) std::__1::__invoke<void (std::__1::__async_assoc_state<void, std::__1::__async_func<vcpkg::cmd_execute_and_capture_output_parallel(vcpkg::Span<vcpkg::Command const>, vcpkg::WorkingDirectory const&, vcpkg::Environment const&)::$_2> >::*)(), std::__1::__async_assoc_state<void, std::__1::__async_func<vcpkg::cmd_execute_and_capture_output_parallel(vcpkg::Span<vcpkg::Command const>, vcpkg::WorkingDirectory const&, vcpkg::Environment const&)::$_2> >*, void>(void (std::__1::__async_assoc_state<void, std::__1::__async_func<vcpkg::cmd_execute_and_capture_output_parallel(vcpkg::Span<vcpkg::Command const>, vcpkg::WorkingDirectory const&, vcpkg::Environment const&)::$_2> >::*&&)(), std::__1::__async_assoc_state<void, std::__1::__async_func<vcpkg::cmd_execute_and_capture_output_parallel(vcpkg::Span<vcpkg::Command const>, vcpkg::WorkingDirectory const&, vcpkg::Environment const&)::$_2> >*&&) type_traits:3859
    #10 0x10332c3f0 in void std::__1::__thread_execute<std::__1::unique_ptr<std::__1::__thread_struct, std::__1::default_delete<std::__1::__thread_struct> >, void (std::__1::__async_assoc_state<void, std::__1::__async_func<vcpkg::cmd_execute_and_capture_output_parallel(vcpkg::Span<vcpkg::Command const>, vcpkg::WorkingDirectory const&, vcpkg::Environment const&)::$_2> >::*)(), std::__1::__async_assoc_state<void, std::__1::__async_func<vcpkg::cmd_execute_and_capture_output_parallel(vcpkg::Span<vcpkg::Command const>, vcpkg::WorkingDirectory const&, vcpkg::Environment const&)::$_2> >*, 2ul>(std::__1::tuple<std::__1::unique_ptr<std::__1::__thread_struct, std::__1::default_delete<std::__1::__thread_struct> >, void (std::__1::__async_assoc_state<void, std::__1::__async_func<vcpkg::cmd_execute_and_capture_output_parallel(vcpkg::Span<vcpkg::Command const>, vcpkg::WorkingDirectory const&, vcpkg::Environment const&)::$_2> >::*)(), std::__1::__async_assoc_state<void, std::__1::__async_func<vcpkg::cmd_execute_and_capture_output_parallel(vcpkg::Span<vcpkg::Command const>, vcpkg::WorkingDirectory const&, vcpkg::Environment const&)::$_2> >*>&, std::__1::__tuple_indices<2ul>) thread:287
    #11 0x103329908 in void* std::__1::__thread_proxy<std::__1::tuple<std::__1::unique_ptr<std::__1::__thread_struct, std::__1::default_delete<std::__1::__thread_struct> >, void (std::__1::__async_assoc_state<void, std::__1::__async_func<vcpkg::cmd_execute_and_capture_output_parallel(vcpkg::Span<vcpkg::Command const>, vcpkg::WorkingDirectory const&, vcpkg::Environment const&)::$_2> >::*)(), std::__1::__async_assoc_state<void, std::__1::__async_func<vcpkg::cmd_execute_and_capture_output_parallel(vcpkg::Span<vcpkg::Command const>, vcpkg::WorkingDirectory const&, vcpkg::Environment const&)::$_2> >*> >(void*) thread:298
    #12 0x18ae6a068 in _pthread_start+0x90 (libsystem_pthread.dylib:arm64e+0x7068)
    #13 0x18ae64e28 in thread_start+0x4 (libsystem_pthread.dylib:arm64e+0x1e28)

==7417==Register values:
 x[0] = 0x0000000000000000   x[1] = 0x000000010fff00e0   x[2] = 0x000000016d02c7b8   x[3] = 0x0000000000000021  
 x[4] = 0x0000000000002088   x[5] = 0x0000000000000001   x[6] = 0x000000016cfac000   x[7] = 0x0000000000000001  
 x[8] = 0x00000001389d9280   x[9] = 0x0000000000000001  x[10] = 0x0000000000000410  x[11] = 0x0000000114ac0180  
x[12] = 0x47ae41c86816a996  x[13] = 0x00000000ffffffff  x[14] = 0x0000000000007e01  x[15] = 0x0000000000000006  
x[16] = 0x000000018ae64e9c  x[17] = 0x0000000107e24720  x[18] = 0x0000000000000000  x[19] = 0x00000001e6536dd0  
x[20] = 0x0000000105577cc0  x[21] = 0x000000010ec23a40  x[22] = 0x00000001105152e0  x[23] = 0x000000000000000f  
x[24] = 0x00000001e4025000  x[25] = 0x0000019d0000387f  x[26] = 0x0000000000000000  x[27] = 0x0000000000000000  
x[28] = 0x0000000000000000     fp = 0x000000016d02d0d0     lr = 0x000000018ad7c2c8     sp = 0x000000016d02d050  
AddressSanitizer can not provide additional info.
SUMMARY: AddressSanitizer: BUS (libsystem_c.dylib:arm64e+0x522bc) in popen+0x230
Thread T407 created by T0 here:
    #0 0x107dd0c5c in wrap_pthread_create+0x54 (libclang_rt.asan_osx_dynamic.dylib:arm64e+0x38c5c)
    #1 0x1033296cc in std::__1::__libcpp_thread_create(_opaque_pthread_t**, void* (*)(void*), void*) __threading_support:421
    #2 0x103328dd0 in std::__1::thread::thread<void (std::__1::__async_assoc_state<void, std::__1::__async_func<vcpkg::cmd_execute_and_capture_output_parallel(vcpkg::Span<vcpkg::Command const>, vcpkg::WorkingDirectory const&, vcpkg::Environment const&)::$_2> >::*)(), std::__1::__async_assoc_state<void, std::__1::__async_func<vcpkg::cmd_execute_and_capture_output_parallel(vcpkg::Span<vcpkg::Command const>, vcpkg::WorkingDirectory const&, vcpkg::Environment const&)::$_2> >*, void>(void (std::__1::__async_assoc_state<void, std::__1::__async_func<vcpkg::cmd_execute_and_capture_output_parallel(vcpkg::Span<vcpkg::Command const>, vcpkg::WorkingDirectory const&, vcpkg::Environment const&)::$_2> >::*&&)(), std::__1::__async_assoc_state<void, std::__1::__async_func<vcpkg::cmd_execute_and_capture_output_parallel(vcpkg::Span<vcpkg::Command const>, vcpkg::WorkingDirectory const&, vcpkg::Environment const&)::$_2> >*&&) thread:314
    #3 0x103326b48 in std::__1::thread::thread<void (std::__1::__async_assoc_state<void, std::__1::__async_func<vcpkg::cmd_execute_and_capture_output_parallel(vcpkg::Span<vcpkg::Command const>, vcpkg::WorkingDirectory const&, vcpkg::Environment const&)::$_2> >::*)(), std::__1::__async_assoc_state<void, std::__1::__async_func<vcpkg::cmd_execute_and_capture_output_parallel(vcpkg::Span<vcpkg::Command const>, vcpkg::WorkingDirectory const&, vcpkg::Environment const&)::$_2> >*, void>(void (std::__1::__async_assoc_state<void, std::__1::__async_func<vcpkg::cmd_execute_and_capture_output_parallel(vcpkg::Span<vcpkg::Command const>, vcpkg::WorkingDirectory const&, vcpkg::Environment const&)::$_2> >::*&&)(), std::__1::__async_assoc_state<void, std::__1::__async_func<vcpkg::cmd_execute_and_capture_output_parallel(vcpkg::Span<vcpkg::Command const>, vcpkg::WorkingDirectory const&, vcpkg::Environment const&)::$_2> >*&&) thread:306
    #4 0x103326090 in std::__1::future<void> std::__1::__make_async_assoc_state<void, std::__1::__async_func<vcpkg::cmd_execute_and_capture_output_parallel(vcpkg::Span<vcpkg::Command const>, vcpkg::WorkingDirectory const&, vcpkg::Environment const&)::$_2> >(std::__1::__async_func<vcpkg::cmd_execute_and_capture_output_parallel(vcpkg::Span<vcpkg::Command const>, vcpkg::WorkingDirectory const&, vcpkg::Environment const&)::$_2>&&) future:2159
    #5 0x10330144c in std::__1::future<std::__1::__invoke_of<std::__1::decay<vcpkg::cmd_execute_and_capture_output_parallel(vcpkg::Span<vcpkg::Command const>, vcpkg::WorkingDirectory const&, vcpkg::Environment const&)::$_2&>::type>::type> std::__1::async<vcpkg::cmd_execute_and_capture_output_parallel(vcpkg::Span<vcpkg::Command const>, vcpkg::WorkingDirectory const&, vcpkg::Environment const&)::$_2&>(std::__1::launch, vcpkg::cmd_execute_and_capture_output_parallel(vcpkg::Span<vcpkg::Command const>, vcpkg::WorkingDirectory const&, vcpkg::Environment const&)::$_2&) future:2210
    #6 0x1032ff664 in vcpkg::cmd_execute_and_capture_output_parallel(vcpkg::Span<vcpkg::Command const>, vcpkg::WorkingDirectory const&, vcpkg::Environment const&) system.process.cpp:496
    #7 0x103377618 in vcpkg::decompress_in_parallel(vcpkg::Span<vcpkg::Command const>) archives.cpp:334
    #8 0x103468f8c in (anonymous namespace)::ArchivesBinaryProvider::try_restore_n(vcpkg::Span<vcpkg::InstallPlanAction const* const>, vcpkg::Path const&) const binarycaching.cpp:258
    #9 0x103465d00 in (anonymous namespace)::ArchivesBinaryProvider::prefetch(vcpkg::Span<vcpkg::InstallPlanAction const>, vcpkg::Span<vcpkg::CacheStatus* const>) const binarycaching.cpp:215
    #10 0x1033977cc in vcpkg::BinaryCache::prefetch(vcpkg::Span<vcpkg::InstallPlanAction const>) binarycaching.cpp:1421
    #11 0x104202300 in vcpkg::Install::perform(vcpkg::VcpkgCmdArguments const&, vcpkg::ActionPlan&, vcpkg::KeepGoing, vcpkg::VcpkgPaths const&, vcpkg::StatusParagraphs&, vcpkg::BinaryCache&, vcpkg::IBuildLogsRecorder const&, vcpkg::CMakeVars::CMakeVarProvider const&) install.cpp:536
    #12 0x1039399a0 in vcpkg::Commands::CI::perform_and_exit(vcpkg::VcpkgCmdArguments const&, vcpkg::VcpkgPaths const&, vcpkg::Triplet, vcpkg::Triplet) commands.ci.cpp:591
    #13 0x103950f00 in vcpkg::Commands::CI::CICommand::perform_and_exit(vcpkg::VcpkgCmdArguments const&, vcpkg::VcpkgPaths const&, vcpkg::Triplet, vcpkg::Triplet) const commands.ci.cpp:670
    #14 0x102eed49c in (anonymous namespace)::inner(vcpkg::Filesystem&, vcpkg::VcpkgCmdArguments const&) vcpkg.cpp:143
    #15 0x102ee9fa0 in main vcpkg.cpp:350
    #16 0x18ab3fe4c  (<unknown module>)

==7417==ABORTING
[1]    7417 abort      vcpkg ci --x-abi-tools-use-exact-versions
```